spec/machine/AARCH64/MachineOps.thy

```diff
@@ -194,21 +184,25 @@ definition setFPUState :: "fpu_state \<Rightarrow> unit user_monad" where
 \<comment> \<open>FIXME FPU: The underlying register is saved and restored when switching between VCPUs, so this
      will probably need to be moved out of machine\_rest\_monad.\<close>
 consts' enableFpuEL01_impl :: "unit machine_rest_monad"
+consts' enableFpuEL01_val :: "machine_word \<Rightarrow> machine_word"
 definition enableFpuEL01 :: "unit machine_monad" where
-  "enableFpuEL01 \<equiv> machine_op_lift enableFpuEL01_impl"
+  "enableFpuEL01 \<equiv> do
+     modify (vcpu_state_update (vcpu_regs_update (\<lambda>r. r(VCPURegCPACR := enableFpuEL01_val (r VCPURegCPACR)))));
+     machine_op_lift enableFpuEL01_impl
+   od"
 
-consts' readFpuState_val :: "machine_state_rest \<Rightarrow> fpu_state"
 definition readFpuState :: "fpu_state machine_monad" where
   "readFpuState \<equiv>  do
      state_assert fpu_enabled;
-     machine_rest_lift $ gets readFpuState_val
+     gets fpu_state
    od"
 
 consts' writeFpuState_impl :: "fpu_state \<Rightarrow> unit machine_rest_monad"
 definition writeFpuState :: "fpu_state \<Rightarrow> unit machine_monad" where
-  "writeFpuState val \<equiv>  do
+  "writeFpuState fpu \<equiv>  do
      state_assert fpu_enabled;
-     machine_op_lift $ writeFpuState_impl val
+     modify (fpu_state_update (\<lambda>_. fpu));
+     machine_op_lift $ writeFpuState_impl fpu
```

Comment:
Do we still need `writeFpuState_impl`? It's not clear to me what else `writeFpuState` would be doing.

Reply:
Not sure whether the write to machine state has side effects, so have left it in. Happy to remove if we know it won't trigger anything

Reply:
It's the weaker assumption, i.e. overall better if we don't need to assume no side effect. If the proofs work, I'd leave it as is.